### ISSUE TITLE
issue with boot2docker.iso

### ISSUE THREAD

**Issue by P3t3rT** [Parallels/docker-machine-parallels#72]
After installing a new vm, docker cannot connect to the frontend. See attachment
![screen shot 2018-02-02 at 09 10 01](https://user-images.githubusercontent.com/2012178/35858237-02f30b46-0b3c-11e8-938a-f94a390589e6.png)

**Comment by P3t3rT** [comment]
Hi, @legal90 . The problem still exists with the latest versions. I can create the vm, but after stopping the new vm, it will not start again. See attachment.
The debug.log can be found here: https://gist.github.com/P3t3rT/118827450232f17bc31def00aa9876fe

Thanks,
Peter

![screen shot 2018-02-11 at 14 45 40](https://user-images.githubusercontent.com/2012178/36074181-b28328a8-0f3c-11e8-8bfa-49411e6832f0.png)

**Comment by wsantos** [comment]
@P3t3rT I've the same issue, you just need to do a `docker-machine ssh` and press enter multiple times, and it will work, I couldn't find the cause.

**Comment by legal90** [comment]
Hi all,
I've investigated this issue and can confirm that it actually exists and could be reproduced on Parallels driver with all boot2docker iso releases starting since `17.12.0-ce-rc1`. It is very weird, but there is definitely something bad happens in running Docker daemon `17.12.0-ce-rc1` on boot2docker under Parallels Desktop.  

**TL;DR:** Initial VM creation via `docker-machine -d parallels <...>` works absolutely fine. But when we start/restart the VM afterwards, the docker daemon starts very slowly there. `docker-machine` reaches 30 seconds of the provisioning timeout (built-in to docker-machine: [[link]](https://github.com/docker/machine/blob/e0e4da5ac71b78a97b711d6a6c1e6e542e1ca1ed/libmachine/provision/utils.go#L269)). 
In 1-2 minutes after that, the docker daemon starts there and opens the TCP socket successfully, so such commands as `docker-machine ssh` and `docker-machine env` start working fine.

Here are my findings in details:

### 1. Added a debugging echo
I've added the following line to the docker's init file in order to catch the moment when the `dockerd` (the previous line) is actually started and the init script exits: https://github.com/legal90/boot2docker/commit/1fb3c8aedd019dd1074c8743be8a48a8b3047cf5

### 2. Issue exist only on subsequent VM start
The issue _does not_ exist on the initial start on the docker daemon. That's what we have in `/var/log/docker.log` right after the VM creation with `docker-machine create` command:
```
------------------------
/usr/local/bin/dockerd -g "/var/lib/docker" -H unix:// -H tcp://0.0.0.0:2376
--label provider=parallels

 --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs >> "
/var/lib/boot2docker/docker.log"
------------------------DOCKER DAEMON STARTED: Sat Feb 24 15:23:56 UTC 2018
time="2018-02-24T15:23:56Z" level=warning msg="The \"-g / --graph\" flag is deprecated. Please use \"--data-root\" instead"
time="2018-02-24T15:23:56.120887247Z" level=info msg="libcontainerd: started new docker-containerd process" pid=2794
```
We see that Docker daemon process was started at `15:23:56` and first log messages from it appeared at the same moment, at `15:23:56`. There is literally 0 seconds delay.

However, if we run `docker-machine restart` command, it will fail as int he first message. After waiting for some seconds and successful login, we can see the log:
```
------------------------
/usr/local/bin/dockerd -g "/var/lib/docker" -H unix:// -H tcp://0.0.0.0:2376
--label provider=parallels

 --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs >> "
/var/lib/boot2docker/docker.log"
=====> DOCKER DAEMON STARTED: Sat Feb 24 16:15:30 UTC 2018
time="2018-02-24T16:17:56Z" level=warning msg="The \"-g / --graph\" flag is deprecated. Please use \"--data-root\" instead"
time="2018-02-24T16:17:56.822300912Z" level=info msg="libcontainerd: started new docker-containerd process" pid=2157
```
Here we see that docker daemon was started (init script exited) at `16:15:30`, but the actual activity from daemon (when it opens the socket and start sending messages to the log) was only at `16:17:56`, which is 2m26s later!

It is constantly reproducible. 
Attempt n.2 (1m29s of delay.):
```
------------------------
/usr/local/bin/dockerd -g "/var/lib/docker" -H unix:// -H tcp://0.0.0.0:2376
--label provider=parallels

 --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs >> "
/var/lib/boot2docker/docker.log"
=====> DOCKER DAEMON STARTED: Sat Feb 24 16:19:40 UTC 2018
time="2018-02-24T16:21:09Z" level=warning msg="The \"-g / --graph\" flag is deprecated. Please use \"--data-root\" instead"
time="2018-02-24T16:21:09.568641001Z" level=info msg="libcontainerd: started new docker-containerd process" pid=2221
```

Attempt n.3 (1m04s of delay):
```
------------------------
/usr/local/bin/dockerd -g "/var/lib/docker" -H unix:// -H tcp://0.0.0.0:2376
--label provider=parallels

 --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs >> "
/var/lib/boot2docker/docker.log"
=====> DOCKER DAEMON STARTED: Sat Feb 24 16:22:21 UTC 2018
time="2018-02-24T16:23:25Z" level=warning msg="The \"-g / --graph\" flag is deprecated. Please use \"--data-root\" instead"
time="2018-02-24T16:23:25.305477028Z" level=info msg="libcontainerd: started new docker-containerd process" pid=2154
```

### 3. Issue exists only with docker >= `17.12.0-ce-rc1`
I've rebuilt the `boot2docker.iso` several times trying catch the revision when the issue appears. I've  found the it happens after this commit: https://github.com/boot2docker/boot2docker/commit/f996284ecf003356e6d749394e95804dc62708e8

All previous versions, including the parallel branch with v17.09.* releases works fine - the daemon always starts immediately.

### 4. The issue exists only on Parallels Desktop
I checked both `vmwarefusion` and `virtualbox` drivers and there is no such behavior. New docker daemon versions always start fine there.

### Reproduction environment:
- OS X 10.13.3
- Parallels Desktop for Mac `13.2.0-43213` and `13.3.0-43321`. Both are affected
- Docker Machine `v0.13.0`
- docker-machine-driver-parallels `v1.3.0`
- boot2docker:
  - _Affected:_`v18.03.0-ce-rc1`, `v18.02.0-ce`, `v17.12.0-ce-rc2`, `v17.12.0-ce-rc1`
  - _Not affected_: `v17.11.0-ce`, `v17.11.0-ce-rc1` `v17.10.0-ce`,`v17.09.1-ce`


@tianon, do you have any ideas what could be a reason of this, or any suggestions where to look at?

**Comment by legal90** [comment]
Pinging @thaJeztah, @tianon. Do you have any suggestion what could cause this behavior? (see my previous comment).
Maybe there is a way to run `strace` in boot2docker, so we can try to catch what `docker` daemon is actually doing while loading? Any help is appreciated.

**Comment by thaJeztah** [comment]
If it’s Docker 17.11 and up, my primary suspect would be the switch to containerd 1.0.x, but no direct suggestion as to what’s the exact cause

**Comment by legal90** [comment]
Hi guys,
I can confirm that the issue is fixed in [`v18.04.0-ce-rc1`](https://github.com/boot2docker/boot2docker/releases/tag/v18.04.0-ce-rc1)

I can suggest that it might be related to Parallels Tools update to v13.2.0 and Linux kernel update to v4.9.*
